cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaSpecChecker.java

```diff
@@ -0,0 +1,139 @@
+ * warnings and share best practice. The intent is this class will generate warnings about
+ * configurations that aren't necessarily illegal or invalid, but that could potentially
+ * lead to problems.
+ */
+public class KafkaSpecChecker {
+
+    private KafkaSpec spec;
+    private KafkaCluster kafkaCluster;
+    private ZookeeperCluster zkCluster;
+    private String timestamp;
+
+    /**
+     * @param spec The spec requested by the user in the CR
+     * @param kafkaCluster The model generated based on the spec. This is requested so that default
+     *                     values not included in the spec can be taken into account, without needing
+     *                     this class to include awareness of what defaults are applied.
+     * @param zkCluster The model generated based on the spec. This is requested so that default
+     *                     values not included in the spec can be taken into account, without needing
+     *                     this class to include awareness of what defaults are applied.
+     */
+    public KafkaSpecChecker(KafkaSpec spec, KafkaCluster kafkaCluster, ZookeeperCluster zkCluster) {
+        this.spec = spec;
+        this.kafkaCluster = kafkaCluster;
+        this.zkCluster = zkCluster;
+        this.timestamp = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ").format(new Date());
+    }
+
+
+    public List<Condition> run() {
+        List<Condition> notifications = new ArrayList<>();
+        checkKafkaLogMessageFormatVersion(notifications);
+        checkKafkaStorage(notifications);
+        checkZooKeeperStorage(notifications);
+        checkZooKeeperReplicas(notifications);
+        return notifications;
+    }
+
+    private final static Pattern VERSION_REGEX = Pattern.compile("(\\d\\.\\d+).*");
+
+    /**
+     * Checks if the version of the Kafka brokers matches any custom log.message.format.version config.
+     *
+     * Updating this is the final step in upgrading Kafka version, so if this doesn't match it is possibly an
+     * indication that a user has updated their Kafka cluster and is unaware that they also should update
+     * their format version to match.
+     *
+     * @param warnings List to add a warning to, if appropriate.
+     */
+    private void checkKafkaLogMessageFormatVersion(List<Condition> warnings) {
+        String logMsgFormatVersion = kafkaCluster.getConfiguration().getConfigOption(KafkaConfiguration.LOG_MESSAGE_FORMAT_VERSION);
+        String kafkaBrokerVersion = spec.getKafka().getVersion();
+        if (logMsgFormatVersion != null && kafkaBrokerVersion != null) {
+            Matcher m = VERSION_REGEX.matcher(logMsgFormatVersion);
+            if (m.find() && !kafkaBrokerVersion.startsWith(m.group(0))) {
+                warnings.add(buildCondition("KafkaLogMessageFormatVersion",
+                                            "log.message.format.version does not match the Kafka cluster version, which suggests that an upgrade is incomplete."));
```

Comment:
I'm not completely sure about this. Unlike the other checks in this class, I think there are valid use cases where someone want might to stick with the old `log.message.format.version` (e.g. because they're using older clients and can't upgrade them right now). It would be kind of annoying if they're doing that intentionally and they get this warning which they can't remove. But I'm not sure it's worth an annotation to be able to disable the warning, for example.

Reply:
It's a fair point - if the user is aware of the issue and is doing it their way intentionally, the warning is potentially annoying. 

I've come across more people who updated their brokers (albeit not using Strimzi) and either forgot or didn't realise they needed to update the config they'd been using since they started, than I have people who are intentionally using a previous/legacy format. So on balance I think the potential benefit outweighs the cost. 

But this is absolutely an anecdotal and not-very-scientific or representative view :-)